File: benches/analyze.rs
use std::collections::HashMap;
use std::env;
use std::fs;
use std::ops::Not;

use anyhow::Context;
use anyhow::Result;
use ordered_float::OrderedFloat;
use serde_json::Value;
use simd_benches::map_collect;
use tabled::Style;
use tabled::Table;

fn default<T: Default>() -> T {
    T::default()
}

#[derive(Debug)]
struct Data {
    function: String,
    crate_: String,
    variant: String,
    case: String,
    time: f64,
}

fn read_messages(path: &str) -> Result<Vec<Value>> {
    let content = fs::read_to_string(path).with_context(|| format!("path = {path}"))?;
    let mut ans = Vec::new();
    for line in content.lines() {
        let msg: Value = serde_json::from_str(line)?;
        ans.push(msg);
    }
    Ok(ans)
}

fn extract_data(messages: &[Value]) -> Result<Vec<Data>> {
    let mut ans = Vec::new();
    for msg in messages {
        let reason = msg["reason"].as_str().unwrap();
        if reason != "benchmark-complete" {
            continue;
        }

        let id = msg["id"].as_str().unwrap();
        let parts: Vec<&str> = id.split('/').collect();
        let function = parts[0].to_owned();
        let crate_ = parts[1].to_owned();
        let variant = parts[2].to_owned();
        let case = parts[3].to_owned();

        let time = msg["typical"]["estimate"].as_f64().unwrap();

        let data = Data {
            function,
            crate_,
            variant,
            case,
            time,
        };

        ans.push(data);
    }
    Ok(ans)
}

struct TableData {
    title: String,
    rows: Vec<String>,
    cols: Vec<String>,
    content: Vec<Vec<OrderedFloat<f64>>>,
    is_throughput: bool,
}

fn gather_table_data(data: &[Data]) -> Result<Vec<TableData>> {
    let functions: &[&str] = &[
        "base64-check",
        "base64-decode",
        "base64-encode",
        "hex-check",
        "hex-decode",
        "hex-encode",
        "base32-check",
        "base32-decode",
        "base32-encode",
        "uuid-format",
        "uuid-parse",
    ];

    let mut tables: HashMap<String, TableData> = default();

    for d in data {
        let t = tables.entry(d.function.to_owned()).or_insert_with(|| TableData {
            title: d.function.to_owned(),
            rows: default(),
            cols: default(),
            content: default(),
            is_throughput: true,
        });

        let row = format!("{}/{}", d.crate_, d.variant);
        let col = d.case.clone();

        let content = match d.case.parse::<u64>() {
            Ok(count) => {
                let gi = 1024.0 * 1024.0 * 1024.0;
                count as f64 / d.time * 1e9 / gi // GiB/s
            }
            Err(_) => {
                t.is_throughput = false;
                d.time // ns
            }
        };

        let i = match t.rows.iter().position(|r| r == &row) {
            Some(i) => i,
            None => {
                t.rows.push(row);
                t.content.push(default());
                t.rows.len() - 1
            }
        };

        if t.cols.contains(&col).not() {
            t.cols.push(col.clone());
        }

        t.content[i].push(OrderedFloat(content));
    }

    Ok(functions.iter().filter_map(|f| tables.remove(*f)).collect())
}

fn render_table(table_data: &[TableData]) -> Result<Vec<(String, Table)>> {
    let mut ans: Vec<(String, Table)> = default();

    for t in table_data {
        let title = if t.is_throughput {
            format!("{} (GiB/s)", t.title)
        } else {
            format!("{} (ns)", t.title)
        };

        let mut b: tabled::builder::Builder = default();

        {
            let mut record: Vec<String> = vec![default()];
            record.extend(t.cols.iter().cloned());
            b.add_record(record);
        }

        let bold_row: Vec<usize> = map_collect(0..t.cols.len(), |j| {
            let iter = t.content.iter().map(|r| r[j]).enumerate();
            if t.is_throughput {
                iter.max_by_key(|&(_, f)| f).unwrap().0
            } else {
                iter.min_by_key(|&(_, f)| f).unwrap().0
            }
        });

        for (i, row) in t.rows.iter().enumerate() {
            let mut record: Vec<String> = vec![row.clone()];

            #[allow(clippy::needless_range_loop)]
            for j in 0..t.cols.len() {
                let f = t.content[i][j];
                let needs_bold = i == bold_row[j];
                if needs_bold {
                    record.push(format!("**{:.3}**", f));
                } else {
                    record.push(format!("{:.3}", f));
                }
            }

            b.add_record(record);
        }

        let table = b.build().with(Style::markdown());
        ans.push((title, table));
    }

    Ok(ans)
}

fn main() -> anyhow::Result<()> {
    let path = env::args().nth(1).unwrap();
    let messages = read_messages(&path)?;
    let data = extract_data(&messages)?;
    let table_data = gather_table_data(&data)?;
    for (title, table) in render_table(&table_data)? {
        println!("#### {}", title);
        println!();
        println!("{}", table);
        println!();
    }
    Ok(())
}

File: scripts/analyze.py
#!/usr/bin/python3
from dataclasses import dataclass
from typing import Any, Dict, List
import sys
import json

from tabulate import tabulate

BENCHES = [
    {"name": "base64-check", "metric": "throughput"},
    {"name": "base64-decode", "metric": "throughput"},
    {"name": "base64-encode", "metric": "throughput"},
    {"name": "base64-forgiving-decode", "metric": "throughput"},
    {"name": "hex-check", "metric": "throughput"},
    {"name": "hex-decode", "metric": "throughput"},
    {"name": "hex-encode", "metric": "throughput"},
    {"name": "base32-check", "metric": "throughput"},
    {"name": "base32-decode", "metric": "throughput"},
    {"name": "base32-encode", "metric": "throughput"},
    {"name": "uuid-format", "metric": "latency"},
    {"name": "uuid-parse", "metric": "latency"},
    {"name": "ascii-check", "metric": "throughput"},
]


@dataclass
class BenchResult:
    name: str
    metric: str
    functions: List[str]
    cases: List[str]
    data: Dict[str, Dict[str, float]]


def read_jsonl(path: str):
    with open(path) as f:
        for line in f.readlines():
            line = line.strip()
            if not line:
                continue
            yield json.loads(line)


def convert_criterion_jsonl(messages: List[Any]):
    for msg in messages:
        reason: str = msg["reason"]
        if reason != "benchmark-complete":
            continue

        parts = msg["id"].split("/")
        bench = parts[0]
        crate = parts[1]
        variant = parts[2]
        case = parts[3]

        time: float = msg["typical"]["estimate"]

        yield {
            "bench": bench,
            "crate": crate,
            "variant": variant,
            "case": case,
            "time": time,
        }


def append_if_not_exists(l, x):
    if x not in l:
        l.append(x)


def find(l, f):
    for x in l:
        if f(x):
            return x
    raise Exception()


def gather_results(items: List[Any]) -> List[BenchResult]:
    results: Dict[str, BenchResult] = {}

    for item in items:
        name = item["bench"]
        metric = find(BENCHES, lambda x: x["name"] == name)["metric"]
        r = results.setdefault(name, BenchResult(name, metric, [], [], {}))

        function = f'{item["crate"]}/{item["variant"]}'
        append_if_not_exists(r.functions, function)

        case = item["case"]
        append_if_not_exists(r.cases, case)

        time = item["time"]

        if metric == "throughput":
            count = int(case)
            throughput = count / time * 1e9 / (1 << 30)  # GiB/s
            data = throughput
        elif metric == "latency":
            data = time
        else:
            raise Exception()

        row = r.data.setdefault(function, {})
        row[case] = data

    return list(results.values())


def position(l, f):
    for i, x in enumerate(l):
        if f(x):
            return i
    raise Exception()


def render_markdown(results: List[BenchResult]):
    results.sort(key=lambda x: position(BENCHES, lambda y: y["name"] == x.name))

    for result in results:
        metric2unit = {"throughput": "GiB/s", "latency": "ns"}
        unit = metric2unit[result.metric]

        headers = [""] + result.cases

        table = []
        for function, data in result.data.items():
            row = [function]
            for case in result.cases:
                col = list(result.data[f][case] for f in result.functions)

                if result.metric == "throughput":
                    needs_bold = data[case] == max(col)
                elif result.metric == "latency":
                    needs_bold = data[case] == min(col)
                else:
                    raise Exception()

                if needs_bold:
                    cell = f"**{data[case]:5.3f}**"
                else:
                    cell = f"  {data[case]:5.3f}  "

                row.append(cell)

            table.append(row)

        print(f"#### {result.name} ({unit})")
        print()
        print(tabulate(table, headers, tablefmt="github"))
        print()


if __name__ == "__main__":
    path = sys.argv[1]
    messages = list(read_jsonl(path))
    items = list(convert_criterion_jsonl(messages))
    results = gather_results(items)
    render_markdown(results)
